dbms/src/Interpreters/JoinV2/HashJoinPointerTable.cpp

```diff
@@ -0,0 +1,176 @@
+bool HashJoinPointerTable::buildImpl(
+    JoinBuildWorkerData & wd,
+    std::vector<std::unique_ptr<MultipleRowContainer>> & multi_row_containers,
+    size_t max_build_size)
+{
+    Stopwatch watch;
+    size_t build_size = 0;
+    bool is_end = false;
+    while (true)
+    {
+        RowContainer * container = nullptr;
+        if (wd.build_pointer_table_iter != -1)
+            container = multi_row_containers[wd.build_pointer_table_iter]->getNext();
+        if (container == nullptr)
+        {
+            {
+                std::unique_lock lock(build_scan_table_lock);
+                for (size_t i = 0; i < JOIN_BUILD_PARTITION_COUNT; ++i)
+                {
+                    build_table_index = (build_table_index + i) % JOIN_BUILD_PARTITION_COUNT;
+                    container = multi_row_containers[build_table_index]->getNext();
+                    if (container != nullptr)
+                    {
+                        wd.build_pointer_table_iter = build_table_index;
+                        build_table_index = (build_table_index + 1) % JOIN_BUILD_PARTITION_COUNT;
+                        break;
+                    }
+                }
+            }
+
+            if (container == nullptr)
+            {
+                is_end = true;
+                break;
+            }
+        }
+        size_t size = container->size();
+        build_size += size;
+        for (size_t i = 0; i < size; ++i)
+        {
+            RowPtr row_ptr = container->getRowPtr(i);
+            assert((reinterpret_cast<uintptr_t>(row_ptr) & (ROW_ALIGN - 1)) == 0);
+            if constexpr (tagged_pointer)
+                assert(isRowPtrTagZero(row_ptr));
+
+            size_t hash;
+            if constexpr (std::is_same_v<HashValueType, void>)
+            {
+                hash = container->getHash(i);
+            }
+            else
+            {
+                hash = unalignedLoad<HashValueType>(row_ptr + sizeof(RowPtr));
+            }
+
+            size_t bucket = getBucketNum(hash);
+            auto * old_head = reinterpret_cast<RowPtr>(
+                pointer_table[bucket].exchange(reinterpret_cast<uintptr_t>(row_ptr), std::memory_order_relaxed));
+            if constexpr (tagged_pointer)
+            {
+                UInt16 tag = (hash & ROW_PTR_TAG_MASK) | getRowPtrTag(old_head);
+                pointer_table[bucket].fetch_or(
+                    static_cast<uintptr_t>(tag) << (64 - ROW_PTR_TAG_BITS),
+                    std::memory_order_relaxed);
+                old_head = removeRowPtrTag(old_head);
+            }
+            if (old_head != nullptr)
+                unalignedStore<RowPtr>(row_ptr, old_head);
+        }
+
+        if (build_size >= max_build_size)
+            break;
+    }
+    wd.build_pointer_table_size += build_size;
+    wd.build_pointer_table_time += watch.elapsedMilliseconds();
+    return is_end;
+}
+
+template bool HashJoinPointerTable::build<void>(
```

Comment:
When the hash comparison is required before key comparison, the hash value is saved in the row layout. Otherwise, it is stored in a separate column, in this case, the template parameter is set to `void`.